rest/historical_test_data_routes.go

```diff
@@ -0,0 +1,348 @@
+type historicalTestDataHandler struct {
+	sc data.Connector
+	htdFilterHandler
+}
+
+func (h *historicalTestDataHandler) Factory() gimlet.RouteHandler {
+	return &historicalTestDataHandler{sc: h.sc}
+}
+
+func (h *historicalTestDataHandler) Parse(ctx context.Context, r *http.Request) error {
+	h.filter = model.HistoricalTestDataFilter{Project: gimlet.GetVars(r)["project_id"]}
+
+	err := h.parse(r.URL.Query())
+	if err != nil {
+		return errors.Wrap(err, "invalid query parameters")
+	}
+
+	err = h.filter.Validate()
+	if err != nil {
+		return gimlet.ErrorResponse{
+			Message:    err.Error(),
+			StatusCode: http.StatusBadRequest,
+		}
+	}
+
+	return nil
+}
+
+func (h *historicalTestDataHandler) Run(ctx context.Context) gimlet.Responder {
+	data, err := h.sc.GetHistoricalTestData(ctx, h.filter)
+	if err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "problem to fetching historical test data"))
+	}
+
+	resp := gimlet.NewJSONResponse(nil)
+	requestLimit := h.filter.Limit - 1
+	lastIndex := len(data)
+	if len(data) > requestLimit {
+		lastIndex = requestLimit
+
+		err = resp.SetPages(&gimlet.ResponsePages{
+			Next: &gimlet.Page{
+				Relation:        "next",
+				LimitQueryParam: "limit",
+				KeyQueryParam:   "start_at",
+				BaseURL:         baseURL,
+				Key:             data[requestLimit].StartAtKey(),
+				Limit:           requestLimit,
+			},
+		})
+		if err != nil {
+			return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err,
+				"problem paginating response"))
+		}
+	}
+	if err = resp.AddData(data[:lastIndex]); err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "problem creating response"))
+	}
+
+	return resp
+}
+
+func makeGetHistoricalTestData(sc data.Connector) gimlet.RouteHandler {
+	return &historicalTestDataHandler{sc: sc}
+}
+
+// htdFilterHandler handles parsing the url query and populating the
+// HistoricalTestDataFilter for the request.
+type htdFilterHandler struct {
+	filter model.HistoricalTestDataFilter
+}
+
+// parse parses the query parameter values and fills the struct filter field.
+func (h *htdFilterHandler) parse(vals url.Values) error {
+	var err error
+
+	h.filter.Requesters, err = h.readRequesters(h.readStringList(vals["requesters"]))
+	if err != nil {
+		return gimlet.ErrorResponse{
+			Message:    "invalid requesters value",
+			StatusCode: http.StatusBadRequest,
+		}
+	}
+
+	h.filter.Variants = h.readStringList(vals["variants"])
+
+	h.filter.GroupNumDays, err = h.readInt(vals.Get("group_num_days"), 1, htdAPIMaxGroupNumDays, 1)
+	if err != nil {
+		return gimlet.ErrorResponse{
+			Message:    "invalid group_num_days value",
+			StatusCode: http.StatusBadRequest,
+		}
+	}
+
+	h.filter.StartAt, err = h.readStartAt(vals.Get("start_at"))
+	if err != nil {
+		return err
+	}
+
+	h.filter.GroupBy, err = h.readGroupBy(vals.Get("group_by"))
+	if err != nil {
+		return err
+	}
+
+	h.filter.Limit, err = h.readInt(vals.Get("limit"), 1, htdAPIMaxLimit, htdAPIMaxLimit)
```

Comment:
Is the min supposed to be 1 instead of `htdAPIMaxLimit`?

Reply:
The min is 1, the second to last value is the max and the last value is the default value (which is also the max possible value).